### DISCUSSION TITLE
Returning output from a native function

### DISCUSSION THREAD

**Discussion by vicperdana** [microsoft/semantic-kernel#5986] [Q&A] [answered]
Hi team, 

Is there a way to get the output of native function's plugin. In this example below, I'd like to capture `items` when invoking `kernel`.

I'm using Auto Function Calling but can't find where the output results are stored. I mostly followed the [example](https://github.com/microsoft/semantic-kernel/blob/917b790b663d72b03aa607c134e5c23ad0e83433/dotnet/samples/DocumentationExamples/SerializingPrompts.cs), with the addition of Auto Function Calling.

Would this be a case where a handlebar planner could fulfil the need?

```csharp
[Description("Filter catalog items")]
    public class FilterCatalogItem (CatalogClientChatService catalogClientChatService)
    {
    // need to fix this code to ensure catalogServiceClient exists
        [KernelFunction, Description("Return a list of catalog items filtered by name or description")]
        public async Task<CatalogItemsPage> GetCatalogItems(
            [Description("Name or description of the item to be queried")] string? searchText)
        {
            if (catalogClientChatService is not null)
            {
                var items = await catalogClientChatService.SearchItemsAsync(searchText);
                return items;
            }
            else
            {
                throw new InvalidOperationException("CatalogServiceClient is not available");
            }
        }
    }
```

**Comment by dmytrostruk** [comment]
> I'm using Auto Function Calling but can't find where the output results are stored.

@vicperdana Could you please provide an example where you execute auto function calling? If you are triggering auto function calling with `IChatCompletionService`, you should get a result for each function call in `ChatHistory` object.

**Reply by vicperdana** [reply]
Hi @dmytrostruk here's a snippet of the code.  I am not using `IChatCompletionService` as I am streaming the results instead. 

```csharp
        OpenAIPromptExecutionSettings openAIPromptExecutionSettings = new()
        {
            MaxTokens = 200,
            ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions
        };
       // More code...
       var result = _kernel.InvokeStreamingAsync<StreamingChatMessageContent>(
                  prompts["chat"],
                  new(openAIPromptExecutionSettings)
                  {
                      {"request", message},
                      {"history", string.Join("\n", history.Select(x => x.Role + ": " + x.Content)) }
                  }
                  );
              
        // Stream the response
        string resultMessage = "Aspire Shop: ";
        await foreach (var chunk in result)
        {
            if (chunk.Role.HasValue)
            {
                Console.Write(chunk.Role + " > ");
            }
            resultMessage += chunk;
            Console.Write(chunk);
        }
        
        history.AddUserMessage(message);
        history.AddAssistantMessage(resultMessage);
```

**Reply by dmytrostruk** [reply, marked as answer]
@vicperdana In this case, we do not provide function result, because streaming chunks are returning independently from function calling functionality. So, there are a couple of ways how to get each function result:
1. Instead of using `Kernel`, you can use `IChatCompletionService` (you can get it from `kernel.GetRequiredService<IChatCompletionService>()`) and execute streaming with `GetStreamingChatMessageContentsAsync` method. This method accepts `ChatHistory` object, and this object will be updated with each function result.
2. We are going to introduce filters for automatic function calling. Filters will allow you to intercept each function call, check function name, arguments, metadata and after function call you will be able to get function result. PR for this functionality and samples is here and it's going to be released soon: https://github.com/microsoft/semantic-kernel/pull/5936.

Hope this information is helpful!
